brightener.cpp

```diff
@@ -16,38 +16,30 @@ shared_ptr<Image> BrightenWholeImage(shared_ptr<Image> inputImage, int& attenuat
     // For brightening, we add a certain grayscale (25) to every pixel.
     // While brightening, some pixels may cross the max brightness. They are
     // called 'attenuated' pixels
-    attenuatedPixelCount = 0;
     auto brightenedImage = 
         make_shared<Image>(inputImage->m_rows, inputImage->m_columns,
             [inputImage, &attenuatedPixelCount](uint8_t* initPixels) {
-            for (int x = 0; x < inputImage->m_rows; x++) {
-                for (int y = 0; y < inputImage->m_columns; y++) {
-                    int index = inputImage->pixelIndex(x, y);
-                    initPixels[index] =
-                        brightenPixel(inputImage->getPixel(x, y), 25, attenuatedPixelCount);
-                }
-            }
+                inputImage->scanPixels([inputImage, &attenuatedPixelCount, initPixels](uint8_t inputPixel, uint16_t rows, uint16_t col, uint16_t index) {
+                    initPixels[index] = brightenPixel(inputPixel, 25, attenuatedPixelCount);
+                    });
         });
     return brightenedImage;
 }
 
+// Throws the exception std::invalid_argument if there is a mismatch between rows and column of inputImage and imageToAdd
 shared_ptr<Image> AddBrighteningImage(shared_ptr<Image> inputImage, shared_ptr<Image> imageToAdd,
     int& attenuatedPixelCount) {
-    // Try converting this into an exception, so callers don't always need to check the returned bool
-    // if (imageToAdd->m_rows != m_inputImage->m_rows || imageToAdd->m_columns != m_inputImage->m_columns) {
-    //     return false;
-    // }
-    attenuatedPixelCount = 0;
+    
+     if (imageToAdd->m_rows != inputImage->m_rows || imageToAdd->m_columns != inputImage->m_columns) {
+         throw std::invalid_argument("Invalid arguments received. Mismatch between rows and column of inputImage and imageToAdd.\n");
+     }
     auto brightenedImage =
         make_shared<Image>(inputImage->m_rows, inputImage->m_columns,
             [inputImage, imageToAdd, &attenuatedPixelCount](uint8_t* initPixels) {
-            for (int x = 0; x < inputImage->m_rows; x++) {
-                for (int y = 0; y < inputImage->m_columns; y++) {
-                    int index = inputImage->pixelIndex(x, y);
-                    initPixels[index] =
-                        brightenPixel(inputImage->getPixel(x, y), imageToAdd->getPixel(x, y), attenuatedPixelCount);
-                }
-            }
+                inputImage->scanPixels([inputImage, imageToAdd, &attenuatedPixelCount, initPixels](uint8_t inputPixel, uint16_t rows, uint16_t col, uint16_t index) {
```

Comment:
Nice resolution of duplicates 👍 
Deadly use of lambda inside lambda :)
In case this code looks intimidating, you can make named functions that return functions (the functional equivalent of a 'factory'). In fact what you are doing here is manufacturing functions and giving them to `scanPixels` etc.
Also check what ChatGPT has to say (copy entire contents of this file and ask it to remove duplication). See how it compares with this code.